api/tests/functional-tests/backend/metrics/test_classification.py

```diff
@@ -1368,8 +1380,8 @@ def test__compute_curves(
         },
         ("dog", 0.05, "tn"): {"all": 1, "total": 1},
         ("dog", 0.8, "fn"): {
-            "missed_detections": 1,
-            "misclassifications": 1,
+            "missed_detections": 0,
```

Comment:
From the docs:

> Similarly, a false negative is a misclassification if there is a prediction with the same Label.key as the groundtruth on the datum, but the Label.value is incorrect. Any false negatives that do not meet this criteria are considered to be missed_detections.

My understanding from the docs was that a false-negative `misclassification` was one where the groundtruth and prediction dont match BUT have values. A false-negative `missing_detection` in the above example would be the case where no prediction values exist for that key.

Reply:
TP
- label key, value match
- score > threshold

FP - misclassification
- Label key match
- Label value mismatch
- Score > threshold
FP - hallucination
- Prediction exists, ground truth does not exist.
- Score > threshold

FN - misclassification
- label key, value match
- score < threshold
FN - missing detection
- Groundtruth exists but no prediction

TN
- No ground truth
-] No prediction

Reply:
> false-negative misclassification was one where the groundtruth and prediction labels having matching keys but dont match values

this is why the second datum in the screenshot above is a misclassification: we predicted (animal, cat) when the groundtruth was (animal, dog). 

there are no considered predictions for the last datum since none of the predictions have a score greater than the threshold of 0.8, so they all drop out. this means we're left with one missed detection.

Reply:
A prediction having a score less than the threshold is still a valid prediction though. I dont think that a prediction that didnt make a score threshold should be grouped with the condition where no prediction existed at all.

Reply:
> A prediction having a score less than the threshold is still a valid prediction though

what is the point of the score threshold in that case? 

the score threshold is meant to mean "only consider predictions with a score greater than x to be a valid prediction"

Reply:
The point of a score threshold is to determine whether the prediction is positive vs negative.

Whether that prediction is correct determines its truth (True, False).

Combine these and you get TP, FP, FN and TN. 

The variation of `missing_detection` doesnt really map well to the classification task (as compared to the obj det task) as we enforce the existence of predictions to groundtruths at ingestion time. (See `validate_matching_label_keys`)

This logic also applies to `hallucination` for FP, which, if you look at that test never gets a value counted.

Reply:
I reached out to Matt and I think this is a definition issue. `Missing detection` doesnt make sense for classification. The condition of FN that you are referring to fits something closer to a "no winner" condition. 

Matt suggested "No prediction" and im wondering if "Null Prediction" would make more sense.

How does all this sound to you?

Reply:
even if I agreed with you, you're arguing for a change in existing logic which should be left to a separate PR.

the test isn't bugged AFAIK: it's returning exactly what I intended it to return. this same logic is used in many other PR-curve related tests that I hand-wrote, so the fact that your new code only changes one test and not the others seems like further evidence that this PR is introducing a bug.

> This logic also applies to hallucination for FP, which, if you look at that test never gets a value counted.

please give a specific example (like I did above via screenshot) if you think I'm not applying the logic consistently

Reply:
Im not sure i fully understand the disconnect. 

The issue and source of my changes is that the documentation contradicts the existing implementation of this test. The documentation defines any FN prediction as a misclassification as long as its paired with a groundtruth with matching label key. For the example you posted above this would result in two misclassifications and no missing detections.

Furthering this point, since we enforce that all datums have the same label keys present, there should never be a condition where FN predictions get classified as "missing_detections". This also holds for FP predictions where "hallucinations" can only occur if the label key is not present in the ground truth set.

To return the test to what it was we would need to redefine what is now FN `missing_detection` as a no-winner outcome where no predictions meet the score threshold.

Reply:
the bottom line is that a refactoring PR shouldn't change existing tests, otherwise we can't know if your code is faster because of your tweaks or because you're omitting potentially important operations (thereby introducing bugs). something in what you wrote is failing only this test, but not the dozens of other `DetailedPrecisionRecallCurve` tests that use similar logic; this is indicative of a bug.

>  documentation defines any FN prediction as a misclassification as long as its paired with a groundtruth with matching label key

a FN prediction with a score below the threshold isn't considered a valid prediction, so there is no contradiction. this logic was 100% purposeful; you're welcome to disagree with it in a separate PR / ticket. if you think the docs need to be updated, it'd be helpful to suggest a change at the appropriate doc. 

> there should never be a condition where FN predictions get classified as "missing_detections".

this test is the conical example of how you can have a FN that's a `missed_detection`; I literally wrote it to check the score threshold logic. there's a FP `hallucination` test somewhere in our repo, I believe, that checks the same idea for `hallucination`.  this logic was 100% purposeful, and is what we all agreed to when we merged `DetailedPrecisionRecallCurves`. you can suggest changes if you'd like, but those changes should be in another PR.

Reply:
> the bottom line is that a refactoring PR shouldn't change existing tests,

I dont know what more you want me to say. The documentation does not match the existing implementation. Ill get this test passing and continue working via issue #649.

Reply:
Passing tests in #650. Hallucination is not required to pass any test for classification.